### ISSUE TITLE
Readable runtime error messages via source maps

### ISSUE THREAD

**Issue by chrisranderson** [probmods/webppl#292]
I've talked with Andreas a bit about figuring out how to tackle this, and after looking through the code and the old stack trace branch, here's my best shot for how to get it done:

- have sweet.js generate source maps [here](https://github.com/probmods/webppl/blob/4b8c735ee982a04815230d30e93c1f645fdc71a3/src/main.js#L54).
- when [the code is evaluated](https://github.com/probmods/webppl/blob/4b8c735ee982a04815230d30e93c1f645fdc71a3/src/main.js#L179), catch errors and translate the JS stack traces to WebPPL stack traces using the source map.

**Questions**
- What do I do with the code that you compile for every program? It looks like you combine the user's source with something else (header.wppl?).
- I haven't thought much about converting the error message itself. Is it okay if I work on that after stack traces are converted?
- Should I use something like [stacktrace.js](https://github.com/stacktracejs/stacktrace.js) or some other library for parsing the JS stack trace?

Hopefully I'm on the right track.

**Comment by stuhlmueller** [comment]
To clarify, full stack traces are probably too difficult for now, since trampolining eliminates the stack in the compiled code. Our modest goal should be to locate errors in the original webppl code when runtime errors happen. In other words, we're only looking at the top-most element of the stack trace.

I'll let Paul (@null-a) reply to your approach and questions (tomorrow), since he is the person most familiar with `main.js`.

**Comment by null-a** [comment]
>  here's my best shot for how to get it done

I think you're pretty close! Here's a sketch of what I think we need to do:

**1. Get an annotated AST from sweet.js**

What we get back from `sweet.compile` is an AST, and we need it to be annotated with source code locations. We'll need line number, column number and filename. It looks to me as though the AST is already annotated with line and column information, but not the filename.

The parser which sweet.js uses does have an option to include the filename, but there's no way to pass options from `sweet.compile` through to the parser. I can think of two ways to fix this:

* Add the filename information to the AST in a separate post-processing step. (This is [how the sweet.js parser does it](https://github.com/mozilla/sweet.js/blob/da661ba1eb9b22049f7e2f0eed12d79b4de04780/src/parser.js#L5759) anyway.)
* Patch sweet.js to pass the options through. (This isn't as bad as it sounds as we already maintain a fork.)

Of those, I think the first is definitely the way to go.

(The [stack trace branch does exactly this](https://github.com/probmods/webppl/commit/c5aa43f7541ed42189cf1bd7b8612a6fba266832#diff-1c90ff38a08209f9ebd4d05d1e43358eR82), though it's simpler there as we didn't use sweet.js at the time, so the parser is called directly.)

**2. Ensure locations are passed through transforms**

I don't think we need to worry about this right away, I'm just noting it for future reference.

We'll need the transforms (in `src/transforms`) to pass along the source location information. This did happen in the stack trace branch, but lots has changed since then so it may no longer work.

**3. Generate a source map**

When we generate JS code from the AST, generate a corresponding source map. I think this is just a case of passing the right option to [`generate`](https://github.com/probmods/webppl/blob/277211a7af0db0c8241b1fab2f2315a74f421661/src/main.js#L168).

This happened [here](https://github.com/probmods/webppl/commit/c5aa43f7541ed42189cf1bd7b8612a6fba266832#diff-1c90ff38a08209f9ebd4d05d1e43358eR83) in the stack trace branch.

**4. Catch exceptions, map the source of the error back to its original location**

> when the code is evaluated, catch errors and translate the JS stack traces to WebPPL stack traces using the source map.

Exactly! (Though as Andreas mentioned, we only need to worry about the element on the top of the stack.)

> Questions
> It looks like you combine the user's source with something else (header.wppl?).

You're right, `header.wppl` is included with every program, and it is compiled every time. We also have a simple [package system](http://docs.webppl.org/en/master/packages.html). Any `.wppl` files from loaded packages are included in much the same way as `header.wppl`.

> What do I do with the code that you compile for every program? 

Ideally errors in say `header.wppl` would also be mapped back to their original location. I think the approach I sketched above lends itself to this, but if the first version only worked for the main program that would be absolutely fine!

> I haven't thought much about converting the error message itself. Is it okay if I work on that after stack traces are converted?

I'm not quite sure what you have in mind. Hopefully I've answered this elsewhere, but if not try me again!

> Should I use something like stacktrace.js or some other library for parsing the JS stack trace?

Sure! The biggest problem we run into when using libraries is that it's not always straight-forward to include them in the version of webppl we build for the browser. Don't worry about that yet though, if there's a library which can help get a first version working I'd definitely use it.

Hope that helps!

**Comment by stuhlmueller** [comment]
Looks like a good plan!

> Ideally errors in say header.wppl would also be mapped back to their original location. I think the approach I sketched above lends itself to this, but if the first version only worked for the main program that would be absolutely fine!

This raises an interesting point (that probably shouldn't affect the implementation plan for now). I expect that many errors will in fact happen in header functions. For example, people might use the wrong argument order for `map`. If we only have the direct error location, no stack trace, then we won't be able to show people where in *their* program the error happened. Hopefully we can still produce a somewhat informative error message by showing a few relevant lines from `header.wppl`, but this is an argument to reconsider whether we can somehow recover full stack traces. If this causes overhead, we could record additional trace information only in a debug mode.

**Comment by chrisranderson** [comment]
Okay, I'll start working on that. Thanks for the help!

**Comment by null-a** [comment]
Great! Don't hesitate to ask if you run into anything which doesn't make sense.

**Comment by chrisranderson** [comment]
Maybe bad for here, but it takes a relatively long time to evaluate even the program "3", for example. Is this because of the header functions? Can they be pre-compiled/cached somehow?

**Comment by chrisranderson** [comment]
For adding filename, you mean for each node in the AST it should have
```
loc: {
  ...
  source: 'user-filename.wppl'
  ...
}
```
right? I've got that working. I haven't used wppl really, but I'd imagine you can split programs into multiple files. Not sure how that works here.

I was imagining I would get a stack trace back, but since it's eval'd, I don't get a trace. I didn't think about that. I'm not sure how to send back anything useful with the what I got (below), unless I (cringe cringe cringe) searched the file for something matching text in the error message.

![image](https://cloud.githubusercontent.com/assets/5461398/12533202/a6c6ccc8-c1e5-11e5-9bd3-6f56e6f99d5f.png)

**Comment by null-a** [comment]
> it takes a relatively long time to evaluate even the program "3", for example. Is this because of the header functions? Can they be pre-compiled/cached somehow?

Take a look at #62.

> right?  

That looks right.

> I'd imagine you can split programs into multiple files.

The only way to split programs across multiple files is via the package system.

It might be tricky/tedious to thread filenames from the place where packages are loaded to the place where the code is parsed and annotated with location information. If so, perhaps you can just hard code the file names to begin with.

[`parse`](https://github.com/probmods/webppl/blob/277211a7af0db0c8241b1fab2f2315a74f421661/src/main.js#L53) could become something like:

````
function parse(code, macros, filename) {
  var ast = sweet.compile(code, { readableNames: true, ast: true, modules: macros });
  return annotateAst(ast, filename);
}
````

The main program is parsed [here](https://github.com/probmods/webppl/blob/277211a7af0db0c8241b1fab2f2315a74f421661/src/main.js#L155), and that would become something like `parse(code, extra.macros, 'main.wppl')`. `wppl` files from packages are parsed [here](https://github.com/probmods/webppl/blob/277211a7af0db0c8241b1fab2f2315a74f421661/src/main.js#L59). If you don't load any extra packages this is only called on the contents of `header.wppl`, so again the file name could be hard coded.

> I was imagining I would get a stack trace back, but since it's eval'd, I don't get a trace.

`eval` is wrapped in a `try` / `catch` on the stack trace branch, and it seems to work?

When the program raises an exception, the `stack` property of the caught exception object is a string which looks something like:

````
ReferenceError: notDefinedVariable is not defined
    at a (eval at webppl_eval (path/to/main.js:131:10), <anonymous>:1601:27)
    at _trampoline (eval at webppl_eval (path/to/main.js:131:10), <anonymous>:1609:13)
    at _main (eval at webppl_eval (path/to/main.js:131:10), <anonymous>:1615:9)
    at eval (eval at webppl_eval (path/to/main.js:131:10), <anonymous>:1618:1)
    at eval (native)
etc...
````

I'd guess that `<anonymous>:1601:27` is the source of the error in the evaled program.

It looks like this info is parsed out in [`stacktraceParser.js`](https://github.com/probmods/webppl/blob/c5aa43f7541ed42189cf1bd7b8612a6fba266832/src/stacktraceParser.js).  (This file also contains the code which looks this info up in the source map.) It looks like we can use this largely unchanged?

One possible complication is that this might not work in all browsers. In the example above I was using nodejs, which is where I would try to get this working first.

Does that help?

**Comment by chrisranderson** [comment]
Weird. I can log `exception.stack` but using [this debugging setup](http://docs.webppl.org/en/master/debugging.html), it says the stack trace is undefined.

I have the filename inserted appropriately, and I have it working to some extent. Is this what you're looking for?

![image](https://cloud.githubusercontent.com/assets/5461398/12657250/a600153c-c5bf-11e5-9b68-ada4f9be6ee8.png)

**Comment by null-a** [comment]
> Is this what you're looking for?

Sure is, looks great!

> Did the stack_trace branch do more?

On a program similar to your first example, it does something very similar:

```
examples/foo.wppl:7
return a

ReferenceError: a is not defined
    at a (examples/foo.wppl:7:8)
    at _main (path/to/main.js:131:10)
    at eval (path/to/main.js:131:10)
````

(It also print the JS stack, which I'm not showing here.)

The stack trace branch does work if you'd like to try it. The only thing I needed to do to was `npm install source-map`. Though when I switched back to the dev branch, I had to do `npm install` **twice** to get things working again.

> Andreas mentioned that "we're only looking at the top-most element of the stack trace."

Yes, although see [this comment](https://github.com/probmods/webppl/issues/292#issuecomment-173683512) for why we might want to recreate more of the stack eventually. (Though probably as a separate piece of work to this.)

What does your current implementation do when an error occurs in the header? e.g. Andreas' example of the args to `map` been switched:

````js
map([42], idF) // wrong, args switched
````

As Andreas suggested, we might be able to provide useful information for this case by showing a few extra lines from `header.wppl`.

> Also worked for a TypeError, but not this RangeError:

Part of the problem here is that you're calling a JS function as a webppl function. (i.e. You need to use member expressions to call JS.)

> t is referring to runTrampoline = function (t) ... t = t(), I think.

I think you're right. The trampoline code is added in `src/transforms/trampoline.js`. Perhaps we can add source locations there too, so that we get something more informative than `null:null:null`?

**Comment by chrisranderson** [comment]
Here's what happens when the error is in the header:

![image](https://cloud.githubusercontent.com/assets/5461398/12827403/cf0144c4-cb3c-11e5-9b71-8481d498dec2.png)

`sandbox.wppl` doesn't have near that many lines in it, so that's interesting.

> we might be able to provide useful information for this case by showing a few extra lines from header.wppl.

> Perhaps we can add source locations there too

What do you mean by both of these? How would this help a user trying to debug their code? Also, what should I do next?

**Comment by null-a** [comment]
> sandbox.wppl doesn't have near that many lines in it, so that's interesting.

I think the line/col numbers point to the correct place in `header.wppl`, it's just the filename that's wrong.

>> we might be able to provide useful information for this case by showing a few extra lines from header.wppl.

>> Perhaps we can add source locations there too

>What do you mean by both of these? How would this help a user trying to debug their code?

**Showing extra context**

Take the example in your previous comment. It's hard to figure out what happened from just `TypeError: fn is not a function`. But if we also show the relevant code from `header.wppl` the user will see something like:

````
TypeError: fn is not a function

var map = function(fn, ar) {
  return ar.length === 0 ? [] : [fn(ar[0])].concat(map(fn, ar.slice(1)));
---------------------------------^
};
````

From this, the hope is that they stand a reasonable chance of figuring out that they've passed something which isn't a function as the first argument to `map`. (Somewhere in their program.) This isn't as helpful as recovering the stack of course, but it might help.

**Showing source locations for trampoline errors**

One of examples you posted includes the error `TypeError: t is not a function at null:null:null`. The reason this happened is that there's no source location information in the trampoline code. (See my previous comment for a pointer to where this code comes from.) There's nothing we can do here that's super useful for users, but I think there'd be some value in showing something like `TypeError: t is not a function at <trampoline>:0:0`.

> Also, what should I do next?

Whatever you'd like to do!

If you want to get this to the point where it can be merged I think we need to:

* Show the correct filename for errors coming from the header and from packages. I suspect that this is conceptually simple but that it requires familiarity with the package code. How about l work on making the changes required to thread filenames through to the place where code is parsed? Once that's done, you can pull in those changes and get your part working.
* Show the code surround the source of the error, similar to the example above. (If instead you wanted to work on recovering the stack, then of course that's fine too!)

Also, don't worry about showing the source of errors in the trampoline code for now. If you did make changes, they'd probably need re-doing once #291 is merged anyway.

**Comment by longouyang** [comment]
Pardon my ignorance if this isn't sensible at all but: would switching from continuations to promises buy us anything here?

**Comment by chrisranderson** [comment]
Okay. I'll get the context showing and wait on the header/package filenames until you're done.

**Comment by chrisranderson** [comment]
Okay here is one sample:
![image](https://cloud.githubusercontent.com/assets/5461398/12893730/16777088-ce50-11e5-90c6-1afa3dcdb99b.png)

Inspired by [Elm error messages](http://elm-lang.org/blog/compilers-as-assistants) - not quite as nice, but a step in that direction. Another alternative:
![image](https://cloud.githubusercontent.com/assets/5461398/12893996/92b36f02-ce51-11e5-9ae4-bf6a5f845f5e.png)

Seems like listing the column number isn't all that helpful.

**Comment by null-a** [comment]
> Okay here is one sample:

Cool!

I'll try and make the changes we need to get the correct filenames some time in the next few days.

> Seems like listing the column number isn't all that helpful.

One place they *might* be useful is the browser code editor where we might want to use the information to position the cursor or highlight code etc. I'm not sure how best to set things up to achieve that, but we can probably ignore it for now and worry about it later.

**Comment by null-a** [comment]
> would switching from continuations to promises buy us anything here?

No idea! What do you have in mind?

**Comment by longouyang** [comment]
From what I know, promises address a couple of problems that continuations have: (1) the indentation of doom and (2) error handling.

So instead of writing this:
```javascript
step1(function (value1) {
    step2(value1, function(value2) {
        step3(value2, function(value3) {
            step4(value3, function(value4) {
                // Do something with value4
            });
        });
    });
});
```

You can write this:

```javascript
Q.fcall(promisedStep1)
.then(promisedStep2)
.then(promisedStep3)
.then(promisedStep4)
.then(function (value4) {
    // Do something with value4
})
.catch(function (error) {
    // Handle any error from all above steps
})
.done();
```

(HT documentation for the [Q library](https://github.com/kriskowal/q)

With promises, we might have a debug mode for generating code where we place a `catch` after every `then`, with each `catch` holding the relevant source map information for its associated `then`.

**Comment by null-a** [comment]
It's an interesting idea! Promises as an improvement to the indentation of doom makes sense for anyone crazy enough to write CPS by hand, but I don't know whether it makes sense as a sort of intermediate representation for a compiler. Assuming it *could* be made to work, it seems like it would be a *lot* of work to re-write all our transforms to use this rather than CPS?

**Comment by longouyang** [comment]
> Assuming it could be made to work, it seems like it would be a lot of work to re-write all our transforms to use this rather than CPS?

Ah.. it's probably not worth it then.

**Comment by null-a** [comment]
> How about l work on making the changes required to thread filenames through to the place where code is parsed?

@chrisranderson: This has been merged into the `dev` branch. The [`parse`](https://github.com/probmods/webppl/blob/af1d2fee45a2c02c751a4520419771c55f5f9f53/src/main.js#L53) function now receives the original filename of the code passed to it. See #327 for a little more detail. Hopefully this will allow you to show the correct filename in your error messages.

**Comment by chrisranderson** [comment]
Okay, I'm working on it. Just don't look too closely to my commit history before the final product... :)

**Comment by chrisranderson** [comment]
So it looks like escodegen will allow you to specify the originating code, but only from one file (I commented on [this escodegen issue](https://github.com/estools/escodegen/issues/267)). Here's my hard-coded version that works:

![image](https://cloud.githubusercontent.com/assets/5461398/13237372/c68ebe84-d988-11e5-9d54-c1b8635eabb2.png)
![image](https://cloud.githubusercontent.com/assets/5461398/13237382/cf8080c2-d988-11e5-86bb-de9dc9806ee8.png)

I'm not sure how to generalize: what if the header isn't used, and don't you have a system for splitting your code into multiple .wppl files?

p.s. Could be nice to add an explanatory message if we notice that the error is coming from header.wppl, e.g. "It looks like the error is coming from header.wppl. Make sure you are using built-in wppl functions correctly."

**Comment by null-a** [comment]
Thanks for all the work you've done on this, it looks great.

Let me see if I've understood the situation correctly. It seems there are two related problems here:

* First, in `compile` we need access to the source code of each AST so that it can be bundled into `codeWithMap`. We already have the code for the main program but we don't have the contents of `header.wppl` or any packaged code. Your current workaround is to re-read the header from disk.
* Second, escodegen only allows us to bundle one source file into `codeAndMap`. You're working around this by setting `codeAndMap.map.sourcesContent` directly. This works as long as the header isn't optimized away and no packages are used, but won't otherwise.

Is that right?

If so, for the second part, assuming we can't make encodegen do the right thing, I wonder if we can extend your current approach. It looks like `codeAndMap.map.sources` is already populated with the names of the files included in the final output. Could we continue to populate `codeAndMap.map.sourcesContent` ourselves, but just take care to make sure it matches up with `codeAndMap.map.sources`? (I'm assuming that getting these populated correctly is what is important to make the downstream stuff work.)

Of course, to do this we actually need to get our hands on the original source code, but we can address that by passing the original code around along with the AST after parsing. I have a version of that idea working [here](https://github.com/null-a/webppl/commit/67c6df1b15b24dcd3f3c73d933c5422b3b354461). With that in place we'd have access to an array of sources and a corresponding array of filenames from within `compile`. I think that would be enough to be able to populate `codeAndMap.map.sources` as described above.

Does that sound reasonable? (If so, I'll submit a PR to get my change merged.)

(This all assumes filenames are unique which I'm not convinced will always be the case, but we can probably safely worry about that later.)

**Comment by chrisranderson** [comment]
> First, in `compile`...

Yep.

> Second, escodegen...

Yep.

Yeah if you could get the sources into `compile` that would be awesome! Especially since reading the file doesn't work from anywhere but the webppl directory. I'm excited to get this all cleaned up and working.

**Comment by null-a** [comment]
> Yeah if you could get the sources into compile that would be awesome!

Great. Now #357.

> Especially since reading the file doesn't work from anywhere but the webppl directory.

It even worse than this, because it's not always the case that the source comes from disk. e.g. When we're running in the browser. This should be handled by my change though, so I think we're good.

**Comment by null-a** [comment]
>> Yeah if you could get the sources into compile that would be awesome!

> Great. Now #357.

This has now been merged into dev.

**Comment by chrisranderson** [comment]
As an update, I'll probably be quite busy until the 26th, and then I can continue working on it.

**Comment by null-a** [comment]
Thanks, that's great!

**Comment by chrisranderson** [comment]
I didn't work on it til today, sorry for the wait. It does look like it's working, though:

![image](https://cloud.githubusercontent.com/assets/5461398/14577180/b9f1d0ac-032d-11e6-8b70-6da55ec71d00.png)

**Comment by ngoodman** [comment]
that looks really promising! how can i replicate your experiment? (i.e. is this in dev already? do i need to do any unusual steps in running webppl?)

**Comment by chrisranderson** [comment]
@ngoodman my terrible Git skills destroyed what I did today - I'll hurry and get it back to where it was and push it to my fork.

**Comment by chrisranderson** [comment]
Okay @ngoodman, seems okay now. You can clone from here: https://github.com/chrisranderson/webppl. You can run `sandbox.wppl` and comment the lines in that file to see the two errors.

**Comment by null-a** [comment]
>  It does look like it's working, though

Good stuff! I've had a play, and the examples work very nicely! I've also quickly tried it in the browser, and the examples I tried worked there too which is great. As far as we know then, this is now fully working for errors raised in the main (user supplied) program and the header?

Outside of that, I think I spotted a few things we need to look at:

* `webpplEval` doesn't work anymore. I think the reason is that the return type of `compile` has changed. (We also need to make sure we've updated all other calls to `compile`.)
* Errors raised in back-end JS code aren't reported correctly. e.g. `Enumerate(null)`. To begin with, we could just assume that errors in back-end code can be reported directly without any mapping. (Maybe this should be the default if we don't have a source map?) However, we _do_ apply transforms to JS code, and it might be useful to generate source maps for these too. (We haven't talked about this before, and we can tackle this as a separate piece of work I think.)
* I don't think we'll get friendly errors from code eval'd by `webpplEval`.
* This doesn't work yet for code pulled in from packages, but I think the fix is simple. The error you get when you try is `TypeError: source.split is not a function` which I think stems from the fact that `source` is a buffer not a string. Perhaps specifying utf8 as the encoding for the appropriate call to `readFileSync` in `pkg.js` is the cleanest way to fix this? (See my github profile for some packages, and the [docs](http://docs.webppl.org/en/master/packages.html) for info on how to use them.)

Do you fancy having a go at these?

**Comment by chrisranderson** [comment]
I think I'm beyond the point that I should work on webppl for my job (I work in David Wingate's lab, by the way, I believe @stuhlmueller knows him), so further efforts will be outside of work. I'd still like to help though.

Of what you mentioned, what needs to be fixed so that my code can be merged? I can tackle that first.

**Comment by null-a** [comment]
Great. Getting friendly errors out of `webpplEval` could wait (we'd probably want to make sure that the errors aren't worse than they already are though) but I think we need to take care of the other three.

**Comment by null-a** [comment]
@chrisranderson We're aiming to have this merged by May 16. If you'd like to finish this off then it would be great if you could submit a PR a few days ahead of that so we have chance to review it. There's no problem if you don't have any more time to work on this - I'm happy to grab the commits from your branch and tie up any loose ends myself. Thanks!

**Comment by chrisranderson** [comment]
Looks like it's okay for me to work on it full time, so I'll start doing that. I'll keep you updated on progress.

@stuhlmueller @null-a is anyone going to be representing WebPPL at the PPAML summer school in July-August?

**Comment by stuhlmueller** [comment]
Many of us (including @null-a and me) will be there, teaching webppl during the second week of the summer school. Looking forward to seeing you in real life!

**Comment by null-a** [comment]
> Looks like it's okay for me to work on it full time, so I'll start doing that.

Great. IIRC 2 of the 3 things I suggested we fix are pretty simple one-line fixes (or thereabouts). If it looks like it will take more than a couple hours work to submit a PR let me know, it might be that I've missed/misunderstood something!

**Comment by chrisranderson** [comment]
- Fixed calls to `compile`:
 - `webpplEval` (I suppose - just added `.code` inside the function since compile returns an object now). Errors inside eval are nasty, still.
 - `webppl:compile` called `main:compile`. No pretty errors for compiled files.
 - updated `main:prepare`, not sure how to test it.
- Fixed `Enumerate(null)` issue. If a situation like that rises, it uses the default error handler.

I'll work on the issues with packages.

**Comment by chrisranderson** [comment]
Fixed the issue with package reporting - setting the encoding to utf8 worked great. All tests pass on my end, except for "testPkgMacrosNotAppliedToOtherPkg" in test-macros.js.

**Comment by null-a** [comment]
Thanks - looking good! A few comments:

> If a situation like that rises, it replicates the original error handler, except without showing the actual line. I didn't see how to get that from an Error object.

I wonder whether in this situation `printFriendlyStackTrace` should just return some value to indicate it didn't handle the error, and we can then re-throw the exception in `run`?

> updated main:prepare, not sure how to test it.

That looks fine to me, don't worry about testing it. (The compiler and inference machinery don't depend on the stuff under `src/analysis`.)

Looking at this I spotted one (related) thing we also need to fix. I think these changes break the `generateCode` option accepted by `compile`. When this is set, it looks like we just need to return early with `return transformedAst;`, or something similar.

> No pretty errors for compiled files.

I don't think this is a problem.

> All tests pass on my end

One of the browser tests is failing. (There's some stuff in the docs about running these if you want to run them yourself.) The fix looks simple though - I think there's a call to `compile` (in `browser.js`) that needs `.code` appending.

> except for testPkgMacrosNotAppliedToOtherPkg in test-macros.js

The problem here is that `run` is swallowing the exception that this test is looking out for.

We could change the test to account for this, but thinking about it I think this is one instance of a more general problem.

Another example comes up when webppl is running in the browser. There, it's probably not the case that we'll always want to log the fixed-up error to the console. For example we might want to use the location information to highlight the error in the browser-based editor.

This probably means that `run` should (at most) be concerned with fixing up the source location of exceptions, and not with the detail of how to handle the exception. I'm not sure what the best way to handle this is, but one idea is to re-throw the exception after patching up the stack trace from the source map. This way, we can leave the caller to handle the exception appropriately.

If you'd rather not tackle this as part of this piece of work, perhaps we could instead add an extra option to `run` so that friendly errrors are only shown when requested. For example, we'd enable friendly errors when calling `run` from the file `webppl`, but not when calling it from the macro tests, etc.

**Comment by chrisranderson** [comment]
> I wonder whether in this situation printFriendlyStackTrace should just return some value to indicate it didn't handle the error, and we can then re-throw the exception in run?

I tried re-throwing the exception, but it adds that it was thrown from `printFriendlyStackTrace`, which was weird. That's why I tried re-creating it.

> One of the browser tests is failing.

I changed it to `compile(...).code` in `browser.js`.

> these changes break the generateCode option

I looked at that, but didn't see anywhere the option was being used, so I'm not sure what needs to happen there.

**Comment by longouyang** [comment]
> Another example comes up when webppl is running in the browser. There, it's probably not the case that we'll always want to log the fixed-up error to the console. For example we might want to use the location information to highlight the error in the browser-based editor.

FWIW, the web editor will display (and try to nicely format) the result of any thrown error. So instead of `console.log`ing the error message, throwing an error with the message should work.

**Comment by ngoodman** [comment]
is this (awesome) progress also going to close #314? if so we should close that as a separate issue?

**Comment by chrisranderson** [comment]
Not super sure but I think they are different. This issue only covers the
top of the stack.

**Comment by null-a** [comment]
> Not super sure but I think they are different.

That's correct.

**Comment by chrisranderson** [comment]
@null-a - how should the `generateCode` option work? If it is set, just return the AST only? Also, what should be changed for the browser, given @longouyang's comment?

**Comment by null-a** [comment]
> If it is set, just return the AST only?

Yes. Just doing `return transformedAst` should do it.

> Also, what should be changed for the browser, given @longouyang's comment?

I guess that the general direction of the change needs to be as I suggested earlier when I suggested

>  one idea is to re-throw the exception after patching up the stack trace 

Though if that can't be made to work, we'll have to think of something else. Don't worry if you're unsure, I'll take a look myself next week.

**Comment by chrisranderson** [comment]
For the small things remaining, is it okay if I pass this issue to you? At this point it might be more work for you to guide me through the process than to whip it out real quick.

**Comment by null-a** [comment]
Sure thing, thanks for all your help!

**Comment by chrisranderson** [comment]
Thank you! I hope I haven't stretched your patience too much. Is there anything you'd like me to do with my fork?

**Comment by null-a** [comment]
Not at all, it's been fun!

>  Is there anything you'd like me to do with my fork?

I've pulled your changes into a local branch, but if you could keep it around for another week that might be useful. Other than that, I have everything I need. Thanks again.
